### PR TITLE
 [SPARK-4502] [SQL] Fix reads unnecessary nested fields from Parquet

### PR DESCRIPTION
**Pull request by cenyuhai**
fix issue 4502

### CODE REVIEW

**Comment by cenyuhai** [comment]
fix issue 4502

**Comment by srowen** [comment]
Please explain the change briefly. Also make the title start with SPARK-4502 and include [COMPONENT] for the relevant component.

**Comment by cenyuhai** [comment]
Hi, Owen. I am so sorry for it.This is the first time for me to create a pull request.
When reading a field of a nested column from Parquet, SparkSQL reads and assemble all the fields of that nested column. This pull request is to cut the unnecessary columns.
1. I add a property 'id' for DataType
2. I add a function 'cutUnnecessaryColumns' in ParquetTableScan, this function will cut the Unnecessary Columns in 'output' and return new AttributeReferences.
3.Finally, in function ParquetTypesConverter.convertToString(), I use regular expression to remove the 'id' fields in the json string.

**Comment by OopsOutOfMemory** [comment]
you can do it like this:
`
[SPARK-4502][SQL] Fix reads unnecessary nested fields from Parquet
`

**Comment by cenyuhai** [comment]
Hi, OopsOutOfMemory. Nice to see you here.Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala [hunk 1]
```diff
@@ -53,7 +53,7 @@ object BindReferences extends Logging {
             sys.error(s"Couldn't find $a in ${input.mkString("[", ",", "]")}")
           }
         } else {
-          BoundReference(ordinal, a.dataType, a.nullable)
+          BoundReference(ordinal, input(ordinal).dataType, a.nullable)
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala]: Why this change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala [hunk 2]
```diff
@@ -36,6 +36,8 @@ import org.apache.spark.util.Utils
 
 
 object DataType {
+  private val curId = new java.util.concurrent.atomic.AtomicLong()
+  def newTypeId = curId.getAndIncrement()
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala]: Can you also explain why we need a new id?

**Comment by marmbrus** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala (hunk 2 above)
Comment (cenyuhai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala]: Now we use the AttributeReference to mark a column, but it is not suitable for nested columns, we need to remove some fields in DataType, it is hard to reconstruct the DataType, so I add a new id to uniquely identify the fields.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala (hunk 1 above)
Comment (cenyuhai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala]: before this, we use ' val ordinal = input.indexWhere(_.exprId == a.exprId)' to find the AttributeReference which is equal to 'a', but the dataType in 'a' is compele, the dataType in input(ordinal) has been cutted in file 'ParquetTableOperations', you can see the 'output' property in case class ParquetTableScan

**Comment by marmbrus** [comment]
Hey, thanks for working on this.  I'd love to have this feature, but I'm not sure I understand the implementation as it exists now.  I propose we close this issue for now and discuss the design on the JIRA.
